### DISCUSSION TITLE
Caching until next main request

### DISCUSSION THREAD

**Discussion by jdevinemt** [symfony/symfony#47286] [Q&A] [answered]
In my application, I have quite a few cases where I want to cache certain data, but only until the next main request. Does anyone have suggestions on the best approach(es) for this? 

I'd prefer to keep the data in my redis cache instead in memory via PHP. This is mainly because I have some routes that require a retrieving a lot or pricing and stored rate information and I wan't to reduce their memory usage (in the eyes of PHP). I know that the memory is going to be used either way, but I'd rather have the memory used via redis to keep the PHP script memory usage lower, and not have to increase my max memory limit too much for PHP.

I haven't found any methods in the cache component to mark an item to expire at the end of the request/response lifecycle.

Here are some of the approaches I've considered so far:

1. Store the data in a service's property. This is how most of my legacy code handles this, but it is not ideal as it can lead to high memory usage on certain pages. See second paragraph above for more information on this.
2. Tag all cache items I want to expire after the request/response lifecycle with something like 'garbage-collection'. Then, add an event listener to delete all cache items with this tag to the kernel.response event.
3. Use boolean properties to determine whether certain information has been cached in the current request/response lifecycle yet. If the value is true, call $cache->get(). If the value is false, call $cache->delete() then $cache->get().

Option 2 seems to be my best bet. But I may have missed a better option, or I may just be approaching this the wrong way altogether. Any input would be greatly appreciated!

**EDIT**
See comment thread on selected answer for the full answer.

**Comment by MatTheCat** [comment, marked as answer]
The data you would fetch from Redis will use memory anyway so you’ll end up consuming more. Have you considered the [array cache adapter](https://symfony.com/doc/current/components/cache/adapters/array_cache_adapter.html)?

**Reply by jdevinemt** [reply]
It would use memory, but only until the end of the function where the data is assigned to a local variable ends. Whereas, if I stored it in a property in a service, it would remain in memory for the whole script.

**Reply by MatTheCat** [reply]
You have no guarantee PHP will free memory at the end of the function (and it probably won't) so using a third-party because of this single possibility seems overkill to me.

**Reply by jdevinemt** [reply]
Gotcha. That's good to know. I guess I was just making an assumption based on the fact that PHP handles garbage collection, but I really don't know the intricacies of how it manages that. So, you're right, trying to use Redis to cache something for a single request is probably not a great approach. Thanks for your input.

**Reply by MatTheCat** [reply]
Sorry after reading a bit about PHP memory management it appears I’m completely wrong: variables memory is freed (in the absence of cycles) when they fall out of scope. That being said I don’t know how it would behave taking the Redis extension into account :thinking:

**Reply by jdevinemt** [reply]
Cool. Thanks for the update. If I run into an issue with memory usage on my scripts I'll probably just use the 2nd approach I mentioned in my post where I tag cache items that I want to delete and then delete them in the kernel.response event.

**Reply by stof** [reply]
Be careful in this option 2: you need to use a unique tag name, not `garbage-collection`, so that 2 concurrent requests don't clear each other cache.

**Reply by jdevinemt** [reply]
Good call @stof. I'll keep that in mind.
